File: JavaScript/Co-op/filtering.js
//- Function to simulate 'filtering' operation
function filtering(constraints, S) {
    let inSet = [...S];
    let exSet = [...S];
  
    constraints.forEach(([col, filterFunc]) => {
      const inVals = S.map(row => row[col]).filter(filterFunc);
  
      inSet = inSet.filter(row => inVals.includes(row[col]));
      exSet = exSet.filter(row => !inVals.includes(row[col]));
    });
  
    return [inSet, exSet];
  }
  
//- Function to simulate 'groupby' operation
function groupby(constraints, S) {
    const groups = [];
  
    constraints.forEach(([col, filterFunc]) => {
      const fVals = S.map(row => row[col]).filter(filterFunc);
      const newGroup = S.filter(row => fVals.includes(row[col]));
  
      groups.push(newGroup);
    });
  
    return groups;
  }

//* OPTIONAL FUNCTION TO PRINT DATAFRAME
function printFilteredDataFrame(data, label) {
    console.log(label);
    if (Array.isArray(data) && Array.isArray(data[0])) {
      // Handle nested arrays (for 'groupby' function)
      data.forEach((group, index) => {
        console.log(`Group ${index + 1}:`);
        if (group.length === 0) {
          console.log("Empty DataFrame");
        } else {
          const keys = Object.keys(group[0]);
          console.log(keys.join(" | "));
          console.log("-".repeat(keys.join(" | ").length));
          group.forEach(row => {
            console.log(Object.values(row).join(" | "));
          });
        }
        console.log();
      });
    } else {
      // Handle single array (for 'filtering' function)
      if (data.length === 0) {
        console.log("Empty DataFrame");
      } else {
        const keys = Object.keys(data[0]);
        console.log(keys.join(" | "));
        console.log("-".repeat(keys.join(" | ").length));
        data.forEach(row => {
          console.log(Object.values(row).join(" | "));
        });
      }
    }
    console.log('\n');
  }

File: Python/Co-op/operators/filtering.py
import pandas as pd

#- Function to simulate 'filtering' operation
def filtering(constraints, S):
    in_set = S
    ex_set = S
    for i, val in enumerate(constraints):
        in_vals = list(filter(val[1], S[val[0]].tolist()))
        in_set = in_set.loc[S[val[0]].isin(in_vals)]
        ex_set = ex_set.loc[~S[val[0]].isin(in_vals)]
    return in_set, ex_set

#- Function to simulate 'groupby' operation
def groupby(constraints, S):
    groups = []
    for i, val in enumerate(constraints):
        print(val[0])
        print(val[1])
        print(S[val[0]])
        fvals = list(filter(val[1], S[val[0]].tolist()))
        new_group = S.loc[S[val[0]].isin(fvals)]
        groups.append(new_group)
    return groups

